pkg/types/validation/installconfig.go

```diff
@@ -195,13 +195,15 @@ func validateNetworkingIPVersion(n *types.Networking, p *types.Platform) field.E
 			allErrs = append(allErrs, field.Invalid(field.NewPath("networking", "serviceNetwork"), strings.Join(ipnetworksToStrings(n.ServiceNetwork), ", "), "when installing dual-stack IPv4/IPv6 you must provide two service networks, one for each IP address type"))
 		}
 
-		switch {
-		case p.Azure != nil:
-		case p.BareMetal != nil:
-		case p.None != nil:
-		default:
-			allErrs = append(allErrs, field.Invalid(field.NewPath("networking"), "DualStack", "dual-stack IPv4/IPv6 is not supported for this platform, specify only one type of address"))
+		if os.Getenv("OPENSHIFT_INSTALL_SKIP_IP_FAMILY_PLATFORM_CHECK") != "true" {
```

Comment:
There is a high bar for adding supported environment variables. Unless there is an expectation that this will be used in CI, the few people that would make use of this can build their own installer.

Reply:
it will be used in the CI, the goal is to have an environment that is able to run dual-stack e2e conformance tests, and Azure seems a good candidate, they are already running e2e test upstream in dual-stack https://testgrid.k8s.io/sig-network-dualstack-azure-e2e#dualstack-azure-e2e-master-iptables so we know that that platform supports it.

As a funny note, it removes `OPENSHIFT_INSTALL_AZURE_EMULATE_SINGLESTACK_IPV6` :)

Reply:
This doesn't make sense to me. Either we are supporting dual stack or not. How are we going to run conformance tests for dual stack using a configuration that we do not support? If we *can* pass conformance test, why are we not supporting it?

Reply:
well, to be honest, I just want to be able to use the installer to deploy a dual-stack cluster, if I don't have to use an environment variable to override the validation, much better :smile: , right now the intsaller works for me, it is able to deploy a dual stack cluster in azure, there are some bugs in other components and maybe we'll find some minor things, but overall it works.
I let others decide on the best way to implement it and discuss the "supportability" topic

Reply:
@staebler We officially support dual-stack on bare metal for customers. We would like to allow-but-not-"support" dual-stack on Azure for developers because it's simpler/cheaper than bare metal and should hopefully mostly work. We do not want to officially support dual-stack for customers on Azure in 4.8 because we haven't really put much thought or testing into it yet. (We might officially support it in 4.9.)

My worry is that if we don't block it in the installer, some customer will see "OCP 4.8 supports dual-stack on bare metal" and "OCP 4.8 is based on kubernetes 1.21" and "kubernetes 1.21 supports dual-stack on Azure", and conclude "Therefore OCP 4.8 supports dual-stack on Azure", and then try to install it and have it successfully install but then turn out to be subtly broken later. So the environment variable is to make it so developers/CI can install dual-stack on Azure in 4.8, but customers can't (without at least making it really really obvious that they're doing something totally unsupported).

I made the environment variable platform-agnostic because we may eventually think about dual-stack on AWS/vSphere/OpenStack, etc, too. Although now that I think about it more, if we wanted to do that, we'd probably need code/template changes to the installer as well, so the people testing that out would have to build their own installer, so maybe the environment variable doesn't really help? (Though, again, we might eventually end up in a situation where, eg OCP 4.9 has partially-working vSphere dual-stack support, so we want to allow developers to hack on it, but not let customers install it...)

So... I could leave it as is (with `OPENSHIFT_INSTALL_SKIP_IP_FAMILY_PLATFORM_CHECK`) or change that to be Azure-specific (`OPENSHIFT_INSTALL_ALLOW_DUAL_STACK_AZURE`?) and then if we eventually officially support dual-stack-on-Azure then that environment variable could just go away, but also, we might add ones for other platforms instead (`OPENSHIFT_INSTALL_ALLOW_DUAL_STACK_VSPHERE`, etc).

Or we could allow anyone to bring up a dual-stack Azure cluster, and just rely on customers reading the docs to know they shouldn't do that (hahahaha).

Reply:
I am 100% in favor of blocking the user from using dual-stack in Azure if we are not supporting it. My quibble is only with the environment variable. We either enable a user to do dual-stack or we do not. There is no disabled unless you find this environment variable mode.

Reply:
So you agree we should block customers from using dual-stack in Azure... and you say we can't have an environment variable... so are you saying then that there is no way we can allow developers to enable dual-stack in Azure but not customers?

Reply:
Maybe I am missing something here. Is it burdensome to require a developer that wants to use an unsupported feature to have to build their own installer? I can appreciate that it is easier for said developer to enable the feature via an environment variable. But that pushes the burden of maintenance from the developer wanting to test the feature to the installer team. Everyone wants their own environment variable for their own unique use case, and the installer team cannot support that at scale.

Reply:
it is hard to create automation if you have to build it ...

Reply:
Help me understand what we are dealing with here, because I clearly do not have the context.

Are there existing informing CI jobs that will start to fail if we do not have the environment variable?

Are there CI jobs in another repo that does support IPv6 on Azure even though the OpenShift as a whole does not?

If we want to support IPv6 on Azure in 4.9, can we wait until the master branch opens for 4.9 to enable support for the feature in the installer, without needing an environment variable? Or is there some reason why you need access to IPv6 on Azure in 4.8?

Reply:
So, dual-stack is going to be a thing for baremetal in 4.8 and it is going beta in Kubernetes in 1.21 (that is 4.8 for Openshift) Azure/Microsoft is leading with Redhat the dual-stack Kubernetes effort, and Kubernetes have CI upstream for dualstack in Azure, so we know Azure supports Kubernetes dual-stack.

I don't know exactly how Dual-stack baremetal CI works, but they use series of bash scripts that uses libvirt to create an environment in a super big machine and, until this week, it was running only a small subset of the tests.

In parallel to the baremetal efforts, I've checked the feasibility of using Azure directly and, surprisingly, it worked at first time with only 2 modifications. The CI jobs that I'm setting up for Azure are optional and for self consume, until someone says this has to be "official" supported, but when that happens we'll have everything working and tested.

Reply:
> In parallel to the baremetal efforts, I've checked the feasibility of using Azure directly and, surprisingly, it worked at first time with only 2 modifications. The CI jobs that I'm setting up for Azure are optional and for self consume, until someone says this has to be "official" supported, but when that happens we'll have everything working and tested.

Do you expect to run these optional tests on PRs to ensure that dual-stack continues to work? If it works now, what is keeping us from adding official support, or at least having a target release where we add official support? In the absence of official support, would we feel comfortable telling a customer that we have experimental support? Can a customer try out dual-stack on Azure with the understanding that support is experimental?

Reply:
There are currently no users and no jobs using dual-stack on Azure in OCP, because dual-stack-on-Azure-in-OCP doesn't currently work and never did. However, we are now a few PRs away from having it finally work.

At that point, we want to create `e2e-azure-ovn-dualstack` jobs in a few repos, because the current `e2e-metal-ipi-ovn-dualstack` can't run the full e2e suite because some of the tests don't work on bare metal, but would run correctly on Azure.

Additionally, we'd like random OCP developers to be able to bring up dual stack clusters more easily, so that, eg, when we find a dual-stack-related bug in the console operator, we can assign it to them to fix, rather than having to fix it ourselves (which we mostly have to do now, because most OCP developers do not have any access to bare metal clusters). So we'd like random OCP developers to be able to bring up dual-stack Azure clusters just like they can any other cluster (more or less), and especially, we'd like cluster-bot to be able to bring up dual-stack Azure clusters.

Reply:
OK. I can get behind these reasons for why we want the environment variable. Thank you for walking me to this. Can we rename the environment variable `OPENSHIFT_INSTALL_EXPERIMENTAL_DUAL_STACK` and check that specifically for Azure?